### PR TITLE
Network bootstrap completion indication

### LINKED ISSUE #3274: Add `bootstrap_completed` to small_network
The bool value should be available to other components via a request to the small_network component.  It will be a simple heuristic which is false while the small_network is still acquiring new connections just after being constructed, and will permanently turn true once 2 * gossip_interval has elapsed or the node has symmetric connections to all known addresses

### PR DESCRIPTION
**Pull request by marc-casperlabs**
Closes #3274.

Adds the following features to the networking component:

* A mapping of known addresses to the last seen `NodeId` seen when connecting to said known address is kept (`known_node_ids`).
* A variable `bootstrap_completed` is introduced, which tracks whether the network considers itself "started up". Once set, it can never be cleared.
* The state of `bootstrap_completed` can be queried through a request (`EffectBuilder::get_networking_bootstrap_state`), this is intended to be used with fast sync code.
* Three conditions are introduced, and checked in the networking component (see `update_bootstrapping_state`) to keep the bootstrapping state up to date.

No changelog entry, as this behavior is only visible (sans a log message) with fast sync, which is not present in earlier versions.

### CODE REVIEW

File: resources/local/config.toml [hunk 1]
```diff
@@ -206,6 +206,16 @@ trie_responses = 0
 finalized_approvals_requests = 1
 finalized_approvals_responses = 0
 
+# Threshold for when the networking considers itself sufficiently connected.
+#
+# Some processes will not begin (or fail) unless this threshold is reached.
```
Comment (Fraser999) [inline resources/local/config.toml]:
```suggestion
# Thresholds for considering bootstrapping to the network to be finished.
# 
# When any one of the conditions is met, bootstrapping is considered finished.
# Some processes will not begin (or fail) until bootstrapping is finished.
```

File: resources/local/config.toml [hunk 2]
```diff
@@ -206,6 +206,16 @@ trie_responses = 0
...
+[network.bootstrap_threshold]
```
Comment (Fraser999) [inline resources/local/config.toml]:
```suggestion
[network.bootstrap_thresholds]
```

File: resources/local/config.toml [hunk 3]
```diff
@@ -206,6 +206,16 @@ trie_responses = 0
...
+# A percentage of known addresses has been reached.
+known_address_connection_percentage = 0.5
```
Comment (Fraser999) [inline resources/local/config.toml]: In the config, we already have `saturation_limit_percent` which is an integer, and `acceleration_ftt` which is a percentage defined as a TOML array.  I think we should avoid using a float here, and instead use an integer (i.e. change the `0.5` to `50`.

File: resources/local/config.toml [hunk 4]
```diff
@@ -206,6 +206,16 @@ trie_responses = 0
 finalized_approvals_requests = 1
 finalized_approvals_responses = 0
 
+# Threshold for when the networking considers itself sufficiently connected.
+#
+# Some processes will not begin (or fail) unless this threshold is reached.
+[network.bootstrap_threshold]
+# A percentage of known addresses has been reached.
```
Comment (Fraser999) [inline resources/local/config.toml]:
```suggestion
# Symmetric connections to the given percentage of known addresses have been established.
```

File: node/src/components/small_network.rs [hunk 5]
```diff
@@ -156,6 +156,14 @@ where
     outgoing_manager: OutgoingManager<OutgoingHandle<P>, ConnectionError>,
     /// Tracks whether a connection is symmetric or not.
     connection_symmetries: HashMap<NodeId, ConnectionSymmetry>,
+    /// Mapping of known node addresses to their potentially known `NodeId`.
+    known_node_ids: HashMap<SocketAddr, NodeId>,
+
+    /// An indicator whether or not the networking component considers itself bootstrapped.
+    ///
+    /// Indented to be used by other components (which can request it) that rely on the state of
```
Comment (rafal-ch) [inline node/src/components/small_network.rs]:
```suggestion
    /// Intended to be used by other components (which can request it) that rely on the state of
```

File: node/src/components/small_network.rs [hunk 6]
```diff
@@ -851,6 +886,53 @@ where
...
+            self.bootstrap_completed = true;
+            info!(
+                target = self.cfg.bootstrap_threshold.connection_count,
+                actual = symmetric_connections.len(),
+                "networking bootstrap completed after connection threshold was met"
+            );
+            return;
+        }
+
+        // Check if we have reached the critical of known nodes connected.
```
Comment (rafal-ch) [inline node/src/components/small_network.rs]:
```suggestion
        // Check if we have reached the critical number of known nodes connected.
```

File: node/src/effect.rs [hunk 7]
```diff
@@ -778,6 +778,22 @@ impl<REv> EffectBuilder<REv> {
         .await
     }
 
+    /// Gets the networks current bootstrapping state.
```
Comment (rafal-ch) [inline node/src/effect.rs]:
```suggestion
    /// Gets the network's current bootstrapping state.
```

File: node/src/components/small_network.rs [hunk 8]
```diff
@@ -851,6 +886,53 @@ where
         ret
     }
 
+    /// Updates the bootstrapping state of the node according to available information.
+    fn update_bootstrapping_state(&mut self) {
```
Comment (rafal-ch) [inline node/src/components/small_network.rs:891]: What about adding a unit test to prove that the calculations, filter_maps, and intersections as a whole work as expected?

I'd suggest introducing a function taking the following parameters:
- self.cfg.bootstrap_threshold
- self.connection_symmetries
- self.cfg.known_addresses
- possibly more (compiler will tell)

Then in `update_bootstrapping_state` you can return early on `self.bootstrap_completed` (as you do) or invoke the proposed function.

This will allow adding unit test with different inputs exercising various edge cases.

Also, to have a single source of truth, I'd suggest adding a check for timeout threshold also here, by having yet another testable `if` instead of an async-based `set_timeout` mechanism.

**Comment by marc-casperlabs** [comment]
Like the underlying issue, this is no longer relevant.
